### ISSUE TITLE
Crash with TLS and thread pools

### ISSUE THREAD

**Issue by DavidKinder** [zaphoyd/websocketpp#336]
My application is similar to that described in https://github.com/zaphoyd/websocketpp/issues/293: a thread pool servicing a single io_service that belongs to the application, rather than WebSocket++. Since the changes last year to support thread pools, this works without problems without TLS.

However, when TLS is used, I see occasional crashes, always inside calls to OpenSSL. Reducing the size of the thread pool to 1 makes these go away, so it is reasonably clear that multiple threads are occasionally accessing the same OpenSSL context, causing corruption.

After some debugging, I have tracked down the problem to an omission in the WebSocket++ TLS support. In Boost Asio, async_read() and async_write() are described as **composite operations** (see for example http://www.boost.org/doc/libs/1_53_0/doc/html/boost_asio/reference/async_write/overload1.html) Although the documentation for boost::asio::ssl::stream doesn't make it explicit, I believe that the other asynchronous operations on SSL streams (the initial handshake and the shutdown) also have this property.

Composite operations have the property that they are implemented in terms of multiple calls to the stream object, which will involve intermediate handlers. The boost::asio::ssl::stream documentation notes that "the application must also ensure that all asynchronous operations are performed within the same implicit or explicit strand" but doesn't call out the consequence of this: we need to ensure that any intermediate handlers for composite operations are also called through any strand. Debugging showed that this wasn't happening for WebSocket++ with TLS: these intermediate handlers were just being called on a random thread in the thread pool, leading to disaster.

In order to fix this, we have to provide an implementation of asio_handler_invoke() that does the right thing with strands, and ensure that it is used for all the TLS related stream calls. (See http://www.boost.org/doc/libs/1_55_0/doc/html/boost_asio/reference/asio_handler_invoke.html and, for an example, http://www.boost.org/doc/libs/1_45_0/doc/html/boost_asio/example/invocation/prioritised_handlers.cpp).

The patches below are against the 0.3.0-alpha4 release, and fix the problem for me. They are not quite complete in that a minimal implementation of socket_con_type::wrap_strand() that just returns the handler should be written for the non-TLS case.

Patch for tls.hpp:

```
--- WebSocketPP/websocketpp-drop/websocketpp/transport/asio/security/tls.hpp    2013-10-11 12:39:55.000000000 +0100
+++ ThirdParty/WebSocketPP/include/websocketpp/transport/asio/security/tls.hpp  2014-03-07 13:58:08.776122400 +0000
@@ -50,6 +50,45 @@
 typedef lib::function<lib::shared_ptr<boost::asio::ssl::context>(connection_hdl)>
     tls_init_handler;

+// Wrapper for handlers that exposes the strand, so that intermediate handlers can be invoked through it
+template<typename handler>
+class strand_handler {
+public:
+  typedef lib::shared_ptr<boost::asio::io_service::strand> strand_ptr;
+
+  strand_handler(strand_ptr strand, handler h)
+    : m_strand(strand), m_handler(h) {
+  }
+
+  void operator()() {
+      m_handler();
+  }
+
+  template<typename arg1>
+  void operator()(arg1 a1) {
+      m_handler(a1);
+  }
+
+  template<typename arg1, typename arg2>
+  void operator()(arg1 a1, arg2 a2) {
+      m_handler(a1,a2);
+  }
+
+  void dispatch(lib::function<void()> f) {
+      m_strand->dispatch(f);
+  }
+
+private:
+  strand_ptr  m_strand;
+  handler     m_handler;
+};
+
+// Invoke a handler on the same strand as the wrapped handler
+template<typename function, typename handler>
+void asio_handler_invoke(function f, strand_handler<handler>* h) {
+    h->dispatch(f);
+}
+
 /// TLS enabled Boost ASIO connection socket component
 /**
  * transport::asio::tls_socket::connection implements a secure connection socket
@@ -230,7 +275,7 @@
         // TLS handshake
         m_socket->async_handshake(
             get_handshake_type(),
-            m_strand->wrap(lib::bind(
+            wrap_strand(lib::bind(
                 &type::handle_init,
                 get_shared(),
                 callback,
@@ -272,8 +317,15 @@
     }

     void async_shutdown(socket_shutdown_handler callback) {
-        m_socket->async_shutdown(callback);
+        m_socket->async_shutdown(wrap_strand(callback));
     }
+
+    // Wrap a handler so that intermediate handlers are called on the strand
+    template <typename handler>
+    strand_handler<handler> wrap_strand(handler h) {
+        return strand_handler<handler>(m_strand,h);
+    }
+
 private:
     socket_type::handshake_type get_handshake_type() {
         if (m_is_server) {
```

Patch for connection.hpp:

```
--- WebSocketPP/websocketpp-drop/websocketpp/transport/asio/connection.hpp  2013-12-06 08:18:29.415002400 +0000
+++ ThirdParty/WebSocketPP/include/websocketpp/transport/asio/connection.hpp    2014-03-07 13:55:22.297122400 +0000
@@ -697,7 +827,7 @@
             socket_con_type::get_socket(),
             boost::asio::buffer(buf,len),
             boost::asio::transfer_at_least(num_bytes),
-            m_strand->wrap(lib::bind(
+            socket_con_type::wrap_strand(lib::bind(
                 &type::handle_async_read,
                 get_shared(),
                 handler,
@@ -735,7 +865,7 @@
         boost::asio::async_write(
             socket_con_type::get_socket(),
             m_bufs,
-            m_strand->wrap(lib::bind(
+            socket_con_type::wrap_strand(lib::bind(
                 &type::handle_async_write,
                 get_shared(),
                 handler,
@@ -754,7 +884,7 @@
         boost::asio::async_write(
             socket_con_type::get_socket(),
             m_bufs,
-            m_strand->wrap(lib::bind(
+            socket_con_type::wrap_strand(lib::bind(
                 &type::handle_async_write,
                 get_shared(),
                 handler,
@@ -814,23 +944,23 @@
         timer_ptr shutdown_timer;
         shutdown_timer = set_timer(
             config::timeout_socket_shutdown,
-            lib::bind(
+            m_strand->wrap(lib::bind(
                 &type::handle_async_shutdown_timeout,
                 get_shared(),
                 shutdown_timer,
                 callback,
                 lib::placeholders::_1
-            )
+            ))
         );

         socket_con_type::async_shutdown(
-            lib::bind(
+            m_strand->wrap(lib::bind(
                 &type::handle_async_shutdown,
                 get_shared(),
                 shutdown_timer,
                 callback,
                 lib::placeholders::_1
-            )
+            ))
         );
     }

```

**Comment by jcaden** [comment]
Hi, it's being a while since this bug was reported. I don't know if you have implemented a solution for this but I can confirm that this is still happening in 0.5.1, 0.6.0 and 0.6.0-permessage-deflate.

Have you got any recommendations for fixing this?

**Comment by zaphoyd** [comment]
I don't have one yet, but fixing this is one of the last two release blockers for 0.7.

**Comment by jcaden** [comment]
Ok, many thanks for your fast reply. I'll wait for the 0.7 :)

**Comment by zaphoyd** [comment]
I believe this has been fixed now (latest develop branch / 0.7.0 release). Any further testing from those who were experiencing this on 0.6 would be useful

**Comment by jcaden** [comment]
We have tested it and it is working correctly. Thank you very much.
